### PR TITLE
Add Connection:close header only when needed

### LINKED ISSUE #656: cURL incompatibility with Kaspersky (kesl)
## Summary

I'm facing an issue with some servers having kesl service (Kaspersky) installed and running.

Sometimes wp_remote_* calls fails with error:
```
Scaricamento del pacchetto di installazione da https://downloads.wordpress.org/plugin/jetpack.10.1.zip...
Warning: Il download non è andato a buon fine. "cURL error 18: transfer closed with 332690 bytes remaining to read"
```

The number of bytes is completely randomic.
Since it's not consistent, I've prepared this script to be run over and over (without wp-cli cache of course):

```
while :; do  wp plugin install jetpack && wp plugin delete jetpack && rm -rf /root/.wp-cli/cache/*; done
```

After digging a lot into WpOrg\Requests\Transport\cURL implementation I've found these lines being the cause of the issue:
```php
<?php
// Force closing the connection for old versions of cURL (<7.22).
if (!isset($headers['Connection'])) {
        $headers['Connection'] = 'close';
}
```

Disabling these lines, the issue disappear. Why aren't we checking the cURL version here?
I would then suggest these changes:
1) Add a new const const CURL_7_22_0 = 0x071600
2) Check the version and add that header only if needed

## Your environment

| Environment                                                  | Answer
| ------------------------------------------------------------ | -------
| Operating system and version (e.g. Ubuntu 20.04, CentOS, tried both)
| PHP version                                                  | 7.4 (tried others as well)
| Requests version                                             | provided with WordPress latest
| Autoloader used                                              | WordPress
| Link to your project                                         | into a vpn, not accessible through the internet

## Tested against `develop` branch?
- [ ] I have verified the issue still exists in the `develop` branch of Requests.

### PR DESCRIPTION
**Pull request by mircobabini**
## Pull Request Type

- [x] I have checked there is no other PR open for the same change.

This is a:
- [x] Bug fix
- [ ] New feature
- [ ] Code quality improvement

## Context

Fixes #656.

## Quality assurance

- [ ] This change does NOT contain a breaking change (fix or feature that would cause existing functionality to change).
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added unit tests to accompany this PR.
- [ ] The (new/existing) tests cover this PR 100%.
- [x] I have (manually) tested this code to the best of my abilities.
- [x] My code follows the style guidelines of this project.

### CODE REVIEW

File: src/Transport/Curl.php [hunk 1]
```diff
@@ -363,8 +364,10 @@ private function setup_handle($url, $headers, $data, $options) {
 		$options['hooks']->dispatch('curl.before_request', [&$this->handle]);
 
 		// Force closing the connection for old versions of cURL (<7.22).
-		if (!isset($headers['Connection'])) {
-			$headers['Connection'] = 'close';
+		if ($this->version < self::CURL_7_22_0) {
+			if (!isset($headers['Connection'])) {
```
Comment (jrfnl) [inline src/Transport/Curl.php]:
```suggestion
		if ($this->version < self::CURL_7_22_0 && !isset($headers['Connection'])) {
```

**Review by jrfnl — CHANGES_REQUESTED** [review]
👋🏻 Hi Mirco, thanks for opening the issue and pull request regarding this.

Could a meaningful unit test be added which can proof the issue exists (without the fix) and can safeguard the fix once it is in place ?

**Comment by mircobabini** [comment]
This is complicated. Btw even if I was able to create a unit test to proof the issue exists, it would work on the described environment only.

Is it still useful?

**Comment by mircobabini** [comment]
@jrfnl can you please introduce me to unit tests on this? How do you expect this to be implemented?

**Comment by jrfnl** [comment]
> @jrfnl can you please introduce me to unit tests on this? How do you expect this to be implemented?

@mircobabini I'm not sure what you are asking ?
- if you need to know how to write tests: see the PHPUnit manual
- if you need to know how to write a test for this library, this repo has plenty of tests in the `tests` directory, have a look there for inspiration.
- if you need to know how to write tests for this specific case - please tell us what you are struggling to define for the test/be more specific about what you need help with. My mind reading skills are a bit rusty ;-)

**Comment by jrfnl** [comment]
@mircobabini Do you still want to continue with this PR or should one of us take over ?

**Comment by mircobabini** [comment]
@jrfnl 

> Do you still want to continue with this PR or should one of us take over ?

Thanks for the heads up. I'm unable to replicate this bc I don't have a Kaspersky license anymore.
Btw I think this is something you can work around it with the provided informations and/or some Kaspersky license if you are albe to get one.

If you can/want to take over, please go ahead. Thanks.

**Comment by mircobabini** [comment]
@jrfnl is this in your radar sooner or later?

**Comment by jrfnl** [comment]
> @jrfnl is this in your radar sooner or later?

@mircobabini Well, yes, it is, but it is kind of hard to verify if something is the right fix and actually fixes something if there is no reproduction scenario....

**Comment by jrfnl** [comment]
@mircobabini We've discussed this today and will merge the PR, but the parse error (and CS errors) in the code will need to be fixed first.

File: src/Transport/Curl.php [hunk 2]
```diff
@@ -363,8 +364,9 @@ private function setup_handle($url, $headers, $data, $options) {
 		$options['hooks']->dispatch('curl.before_request', [&$this->handle]);
 
 		// Force closing the connection for old versions of cURL (<7.22).
-		if (!isset($headers['Connection'])) {
-			$headers['Connection'] = 'close';
+		if ($this->version < self::CURL_7_22_0 && !isset($headers['Connection'])) {
+				$headers['Connection'] = 'close';
+			}
```
Comment (jrfnl) [inline src/Transport/Curl.php]:
```suggestion
			$headers['Connection'] = 'close';
```

**Comment by mircobabini** [comment]
> @mircobabini We've discussed this today and will merge the PR, but the parse error (and CS errors) in the code will need to be fixed first.

Nice! I'm sorry I wasn't able to provide tests and a reproduction scenario. But glad to now it will be merged.
Just committed the proposed changes.

**Review by jrfnl — APPROVED** [review]

**Comment by jrfnl** [comment]
Thanks @mircobabini ! The patch will be included in the 2.1.0 release and if there will be a 2.0.x release, we may backport it as well.

**Comment by jrfnl** [comment]
FYI: The patch was included in the Requests 2.0.8 release and reverted in Requests 2.0.9 due to it causing problems with Curl 7.29.0 (and possibly others). See #838

I'll re-open the original issue.

**Comment by mircobabini** [comment]
@jrfnl thanks for sharing this update.

Since this issue is:
- related to Kaspersky which is a security software
- some old CURL versions (probably not secure)
- not so common in general

I'd say it's to be considered as a rare-case backward compatibility patch.
Probably not something we want to deal with - and not something we want to cause main issues to others.

**Pull request merged**